### DISCUSSION TITLE
Logs constantly showing "has ambiguous subject" and "failed removing merged clusters for subject" errors?

### DISCUSSION THREAD

**Discussion by drinn** [photoprism/photoprism#1979] [Q&A]
I just installed photoprism yesterday. Things appear to be working OK. I started playing around with the "People" section today and the interface seems rather slow when trying to correct/add/remove people from photos.

Looking at the logs, I'm seeing constant messages of:

level=warning msg="faces: XICN364AMGJKJX6F4LI6OOIM62GPF47V has ambiguous subject jr67tqm1c9kczt98"
level=warning msg="faces: failed removing merged clusters for subject jr67tqm1c9kczt98"
level=warning msg="faces: YU3BK2KL7M5ZU7CP5NHXZLU7JT3HMIO2 has ambiguous subject jr67tqm1c9kczt98"
level=warning msg="faces: failed removing merged clusters for subject jr67tqm1c9kczt98"

The part preceding the "has ambiguous subject" changes occassionally, but the subject target of "jr67tqm1c9kczt98" always appears to be consistent. I'm not sure if the UI is slow in trying to deal with these errors all the time or what the cause might be? Any ideas?

**Comment by graciousgrey** [comment]
Is your indexer still running?

**Comment by drinn** [comment]
It's not searching for new files in my library if that is what you mean? After the initial index finished, I was trying to correct the faces that it identified incorrectly or not at all. And that's when I noticed photoprism would hang sometimes when modifying entries. Checking the logs during these periods of slowing down is when I noticed those warnings.

**Comment by lastzero** [comment]
Did you try running the "faces audit --fix" command in a terminal yet?

```
docker-compose exec photoprism photoprism faces audit -f
```

**Comment by lastzero** [comment]
As you can see here, this is just a warning that no face embeddings can be removed because very similar faces match different people:

https://github.com/photoprism/photoprism/blob/58a5f940694446958e5cd4b869bea1d2bf9aaf98/internal/query/faces.go#L160


What faces and people could that be? We do not know.

- Maybe you use different names for the same person? This could trigger such a warning.
- Without seeing the actual images and associated data, it's fruitless guesswork and mainly a case for technical support, for which GitHub Issues is the wrong place.

**Reply by drinn** [reply]
Ok. I figured after me trying to help my instance of photoprism along by trying to correct those pictures it got wrong and add more correct people to pictures for better granularity, that these messages would go away after a while. But they've been steadily streaming by all this time. To answer some of your questions:

- I did try doing a "faces audit --fix" after seeing your suggestion, but it didn't do anything to fix the warnings.
- I don't use different names for the same person.
- I'm new to the project and thought this may be a bug, and hence I posted it here. Sorry if this is not the proper channel for this.

At this point, I'd be willing to just wipe all the tagged faces so far and start over but I haven't found a way to do that in the software. Or maybe I can just remove all the storage and the DB itself and start from scratch completely.

In any case, thanks for taking a look at things.

**Reply by lastzero** [reply]
You can use the photoprism faces reset -f command as also shown in the command line examples in the Getting Started docs. If you figure out how to consistently reproduce this warning, we're happy to take a look as it should be rare and not commonly shown all the time.

**Reply by layanto** [reply]
Is it possible for PhotoPrism's log file to list the uid of photos causing the failed removing merged clusters warning? If the issue is due to incorrectly assigning different names for the same person, just using the face id or subject id in the warning/error messages result in too many photos to check which ones are conflicting.

**Reply by lastzero** [reply]
When we implemented the logs, we weren't sure if this would even happen in practice. Now that we know it, we can improve the logs in a future release.

If you know SQL, you can look it up directly in the database using the UID of the subject that is shown in the logs. It's referenced in the `subjects` and `markers` tables:

- `subjects` contains the name of a person
- `markers` [references](https://github.com/photoprism/photoprism/blob/develop/internal/entity/marker.go#L31) `files.file_uid` so you can get the photo UID from the `files` table
- with the `photo_uid`, you can query the `photos` table

**Reply by layanto** [reply]
Is it possible to use SQL to identify which markers are conflicting due to manual subject assignment but automatic face recognition thinks it should be assigned to a different subject? I want to find out if I have any incorrect manual face/subject assignments. I find that after I assigned all unassigned faces, after a few hours, hundreds of photos come back in the faces:new search.

Another question: If a marker of a photo is manually assigned to a subject, will PhotoPrism's conflict resolution potentially un-assign the marker or does it only do that for the automatic assignment (and skip the manual assignments)?

**Reply by lastzero** [reply]
Generally you can debug it with SQL, but it would take me a while to provide you with specific queries and easy to understand instructions. This is difficult with my current workload, which even requires me to work on Sunday evenings, as you can see. As a starting point, compare the `face_id` and `subject_uid` in the faces and markers tables. You can also check our public source code for examples.

There are specific/rare conditions that can cause the assignment of a person to a face to be removed, such as ambiguities within a very small margin of error.

**Reply by layanto** [reply]
Re - "There are specific/rare conditions that can cause the assignment of a person to a face to be removed, such as ambiguities within a very small margin of error.", is this also true for manual assignment? Or does PhotoPrism only remove automatic assignment of a person to a face on ambiguities/conflicts?

**Reply by lastzero** [reply]
https://github.com/photoprism/photoprism/blob/develop/internal/entity/face.go#L167

**Reply by l8gravely** [reply]
I'm coming to this late, but it would be great if there was a command or a way to translate the key given in the warning message, so we could maybe go find that person(s) and look for problems.

I suspect it's because you have kids and pictures from different ages all keyed to the same person.  At least I know I do... and it's just the first thought that popped into my head.

**Reply by ivandbarrera** [reply]
Hi, arriving very late to this conversation. I have two kids, and I take a lot of pictures. I just indexed more than 170k pics, and I see some collisions. I see from the errors that two subjects have conflict with face, so I can usually search for face:<upper case sha> and the subject:<lower case sha> and somewhat gives me an idea. When the conflicts are resolved, the face (upper case SHA) disappears, I assume it merges with the "well known name", which is another "face". 

However, one of those faces is confusing my two kids. And listing a significant number of pictures on the face. I have tried to tag manually, but I can't seem to easily find what is causing the conflict or how can I "hint" to separate the detections so the "face" is resolved. 

Is there an easy guidance to narrow down the conflict images, so I can manually untag/tag and resolve the conflict? Thanks!

Is the "marker" information something I can use to search anything/narrow down?

**Comment by l8gravely** [comment]
>>>>> "Ivan" == Ivan Barrera ***@***.***> writes:

> Hi, arriving very late to this conversation. I have two kids, and I
> take a lot of pictures. I just indexed more than 170k pics, and I
> see some collisions. I see from the errors that two subjects have
> conflict with face, so I can usually search for face: and the
> subject: and somewhat gives me an idea. When the conflicts are
> resolved, the face (upper case SHA) disappears, I assume it merges
> with the "well known name", which is another "face".

> However, one of those faces is confusing my two kids. And listing a
> significant number of pictures on the face. I have tried to tag
> manually, but I can't seem to easily find what is causing the
> conflict or how can I "hint" to separate the detections so the
> "face" is resolved.

> Is there an easy guidance to narrow down the conflict images, so I
> can manually untag/tag and resolve the conflict? Thanks!

I'm not sure I have an answer, but what I did for pictures of my kids
when they were younger is to name them <kid>-<year> when tagging.
Because they changed so much in how they looked over the years as they
grew from 0 to 20.  

And sometimes no matter what I do, it doesn't want to find my wife's
face, or just skips faces.  I wish I could click on a face sometimes
and say "tag this as <person>" no matter what.  

But it usually gets better as things go along.  

Good luck!
